feat(user-test): add Observator role to moderated sessions

Santosh69 opened the pull request:
## Fixes #1222 

### Summary
Implements the **Observator** role for Moderated User Tests, enabling a third participant to view sessions silently and take timestamped notes.

### Functionality
- **Mesh P2P**: Refactored [VideoCall.vue](cci:7://file:///Users/sanhub/RUXAILAB/src/ux/UserTest/components/VideoCall.vue:0:0-0:0) to support multi-peer connections (Moderator + Participant + Observator).
- **Observator Mode**: `recvonly` media config; disabled camera/mic controls; dedicated "Session Notes" drawer.
- **Invitation**: Added role selection (Evaluator/Observator) to `CreateInviteDialog`.


## Verification Checklist

- [x] **Define Observator access level**
  - Added `ACCESS_LEVEL.OBSERVATOR = 3` constant
  - Commits: [`168256a4`](https://github.com/Santosh69/RUXAILAB/commit/168256a4), [`110b8298`](https://github.com/Santosh69/RUXAILAB/commit/110b8298), [`096c261a`](https://github.com/Santosh69/RUXAILAB/commit/096c261a)

- [x] **Add Observator role UI**
  - Role selection in invite dialog with color/icon styling
  - New [ObservatorNotes.vue](cci:7://file:///Users/sanhub/RUXAILAB/src/ux/UserTest/components/ObservatorNotes.vue:0:0-0:0) component for timestamped note-taking
  - Auto-saves notes with task context
  - Commits: [`1bd76097`](https://github.com/Santosh69/RUXAILAB/commit/1bd76097), [`4c86a4c0`](https://github.com/Santosh69/RUXAILAB/commit/4c86a4c0)

- [x] **Enable moderated tests**
  - Made moderated usability testing available in study creation
  - Commit: [`3ae07d6a`](https://github.com/Santosh69/RUXAILAB/commit/3ae07d6a)

- [x] **Observator WebRTC integration**
  - Receive-only P2P peer (no camera/mic sharing)
  - Hidden from video grid, can view all participants
  - Commit: [`168256a4`](https://github.com/Santosh69/RUXAILAB/commit/168256a4)

- [x] **Notes panel in test view**
  - Fixed right-side drawer with timestamped notes
  - Commit: [`168256a4`](https://github.com/Santosh69/RUXAILAB/commit/168256a4)

- [x] **Observer notes timeline UI**
  - Admin timeline view showing observer notes with timestamps
  - Toggle to show/hide observer notes with badge indicator
  - Commits: [`3ffc615b`](https://github.com/Santosh69/RUXAILAB/commit/3ffc615b), [`8b284cb2`](https://github.com/Santosh69/RUXAILAB/commit/8b284cb2)

- [x] **Fix test access and navigation**
  - Improved admin access controls
  - Fixed moderated test redirects
  - Commits: [`751ddd5e`](https://github.com/Santosh69/RUXAILAB/commit/751ddd5e), [`f61c223c`](https://github.com/Santosh69/RUXAILAB/commit/f61c223c)

- [x] **Fix access level checks**
  - Corrected Observator detection logic
  - Fixed `ACCESS_LEVEL` import statements
  - Commits: [`e3eb3ec4`](https://github.com/Santosh69/RUXAILAB/commit/e3eb3ec4), [`15e2dd3a`](https://github.com/Santosh69/RUXAILAB/commit/15e2dd3a), [`8a44aa5e`](https://github.com/Santosh69/RUXAILAB/commit/8a44aa5e)

- [x] **Improve invite dialog**
  - Multi-invite support in CreateInviteDialog
  - Fixed Evaluator role mapping
  - Aligned in-app notifications with email invites
  - Commits: [`acaefcfc`](https://github.com/Santosh69/RUXAILAB/commit/acaefcfc), [`6eb1c113`](https://github.com/Santosh69/RUXAILAB/commit/6eb1c113), [`0af380a2`](https://github.com/Santosh69/RUXAILAB/commit/0af380a2)

- [x] **Integrate SART feature**
  - Merged SART instrument with observer notes
  - Commit: [`2824edb8`](https://github.com/Santosh69/RUXAILAB/commit/2824edb8)

- [x] **Resolve merge conflicts**
  - Integrated upstream changes from develop
  - Resolved conflicts in [useCooperatorUtils.js](cci:7://file:///Users/sanhub/RUXAILAB/src/shared/composables/useCooperatorUtils.js:0:0-0:0) and [VideoCall.vue](cci:7://file:///Users/sanhub/RUXAILAB/src/ux/UserTest/components/VideoCall.vue:0:0-0:0)
  - Commits: [`b71644fa`](https://github.com/Santosh69/RUXAILAB/commit/b71644fa), [`e42b392c
 - [ ] **Notes persistence on refresh**
   - Fix notes loading after page refresh  
   
###    Video Demo 
> Please feel free to skip if slow
   [Observator Role Behaviour](https://drive.google.com/file/d/15CfH_8pyptDEWlQU1lzhMGbbP-ZB2WZS/view?usp=sharing) 
   [Admin-Observator flow](https://drive.google.com/file/d/1rXPms4lAjr1ipvYu1_T1e1URAYproklk/view?usp=sharing)
[Still a Problem](https://drive.google.com/file/d/1vGB2C928rTW2aoasxPh5dOkmSTuXyQjO/view?usp=sharing)

Santosh69 committed:
feat(user-test): implement Observator role with Mesh P2P

Santosh69 commented:
Hiii @marcgc21 , @KarinePistili , @jvJUCA , will also post the video explanation and working..

Santosh69 commented:
Hii @marcgc21 im reviewing this once again.. to ensure works fully for you to review directly..

Santosh69 committed:
fix(cooperators): add Observator role to filter options

marcgc21 committed:
fix(user-test): new role observator must be 3

KarinePistili committed:
fix(merge): resolved merge conflicts

KarinePistili committed:
fix(roles): added observator access level

jvJUCA committed:
fix: deleted const ACCESS_LEVELS

jvJUCA committed:
Merge branch 'feat/user-test-observator-role' of https://github.com/Santosh69/RUXAILAB into feat/user-test-observator-role

KarinePistili committed:
chore: removed incorrect info from cooperators model

marcgc21 committed:
fix(user-test): make available moderated test

marcgc21 committed:
Merge branch 'feat/user-test-observator-role' of https://github.com/santosh69/ruxailab into feat/user-test-observator-role

KarinePistili committed:
fix(videocall): fixed import and onclick function

KarinePistili committed:
fix(cooperators): fixed import error of access_level

jvJUCA committed:
fix: wrong access level to define if user is observator

Santosh69 committed:
fix(merge): resolve useCooperatorUtils conflict by accepting HEAD

Santosh69 committed:
fix(merge): restore useCooperatorUtils and complete merge

Santosh69 commented:
hii @marcgc21, working on this.. will push with explanation soon

Santosh69 commented:
hiii @marcgc21 , quick questions..
**1. should observtor be the participant of the test or can it be anyone as an observator?**

**2. scheduling the test session is optional in moderated tests, right?**  ( im assuming Yes)
**3. shall i implement to invite multiple observators at a time?**

Santosh69 committed:
fix: moderated test access and redirects

Santosh69 committed:
fix(moderated-test): improve admin access and navigation for observer feature

- Fix test structure validation (Object.keys check for object, not array)
- Add admin bypass for 24h window restriction in isStartTestDisabled
- Improve session filtering with proper 24h window calculation
- Fix Preview path to include userId parameter in ManagerView
- Update Start Session route to use correct testview path
- Normalize notification redirect URLs with leading slash

Related to #1222

Santosh69 commented:
Hii @marcgc21 , @KarinePistili, 

> hiii @marcgc21 , quick questions..
>  **1. should observtor be the participant of the test or can it be anyone as an observator?**
> 
> **2. scheduling the test session is optional in moderated tests, right?** ( im assuming Yes)
>  **3. shall i implement to invite multiple observators at a time?**

whenevr you have time to proceed, please answer these and another one:

**4. how does the admin know observator is present or not, shall I implement participants list for this, currently its hardcoded**

WIP , with video explanation 

**Thank you for time!**

marcgc21 commented:
1. Observator is a new role that doesn't exists. Those that are observators they can not be part of the task in terms of answering. They are on the evaluators team, but in that case, they just observe and take notes without interacting with the user
2. Indeed is mandatory, the administrator of the test schedules a meeting and send invites to the other members, other administrators, observators and users.
3. Yeah I think that might be necessary!
4. It might be nice, administrator should now that the observators are online and available

Santosh69 committed:
feat(moderated-test): add observer notes toggle with badge and fix access level bugs

Santosh69 committed:
Merge branch 'develop' into feat/user-test-observator-role

Santosh69 commented:
Hiii @marcgc21 , @KarinePistili , @jvJUCA almost fixed changes locally, pushing commit by commit so that it's cleaner to review,  more commits will follow with proper video demo, actually testing this in evry case ( **multiple observators, admin changes the roles case , task relation to notes**), that's why iam coming across many other issues too.. good thing getting to connect things

**will push changes by EOD, hope its okay**.. 

**also quick question: is it good to have an implementation for admin to send messages(during the test itself) to observator aBout something specific maybe for diff PR, just curious**

Thank you

KarinePistili commented:
> Hiii @marcgc21 , @KarinePistili , @jvJUCA almost fixed changes locally, pushing commit by commit so that it's cleaner to review, more commits will follow with proper video demo, actually testing this in evry case ( **multiple observators, admin changes the roles case , task relation to notes**), that's why iam coming across many other issues too.. good thing getting to connect things
> 
> **will push changes by EOD, hope its okay**..
> 
> **also quick question: is it good to have an implementation for admin to send messages(during the test itself) to observator aBout something specific maybe for diff PR, just curious**
> 
> Thank you

I think for this PR specific no need, but maybe it can be interesting for a next one.

Santosh69 commented:
sure got that @KarinePistili

Santosh69 committed:
feat(admin): add observer notes timeline UI with mock data

Santosh69 committed:
merge: integrate remote SART feature with observer notes implementation

Santosh69 commented:
HI @marcgc21 , i noticed that there is already logic `INviteDialog.vue` inviting mutliple cooperators (used for Heursitics, Card Sorting Unmoderated), and Moderated ones  is following different Invite system, `CreateInviteDialog.vue`
i think because mod tests have scheduling feat and observtor that's why

so  is it okay re-use some logic , is it acceptable? would it be connected?
also `InviteDialog.vue`  is also blocking direct email invite, 

lmk your thoughts, whenevr you have time

antaripdebgupta commented:
Hi @Santosh69,
- *TaskDetailsModal.vue* - `getTaskAnswer` function is empty and unused

Santosh69 commented:
thanks for checking out @antaripdebgupta , i removed it in this: 3ffc615b3daffaf7be33ea9268625754808b0691 in the merge came back, will remove it in next commit

appreciate it!

Santosh69 committed:
Refactor: CreateInviteDialog multi-invite & restore submit; update TaskDetailsModal

Santosh69 commented:
Hii @marcgc21 , @KarinePistili , mostly done and working the Video showcasing all testing, few minor changes to complete PR, 
**1. Participant list to show observator presence for the Admin only maybe
2. Re-checking the direct email invite to the User and  Scheduling aspect **

**Note: i Kept Observator's notes in Admin's view Intentionally HArdcoded, though it  should be handled Separatingly** 

thanks for your time

Santosh69 committed:
Fix: correct Evaluator role mapping in CreateInviteDialog

Santosh69 committed:
fix(invites): align in-app notifications and external email invites

Notify registered users in-app; send Cloud Function emails to external recipients

Santosh69 commented:
fixing the Linting issues.. will be taken care, shall I fix the linting issues here in this branch @marcgc21 @KarinePistili  ,is it okay/easy for you to review?

Santosh69 commented:
hii @marcgc21  shall I handle this [Observator notes Refresh over ride](https://drive.google.com/file/d/1vGB2C928rTW2aoasxPh5dOkmSTuXyQjO/view?usp=sharing) in a new PR or is it okay do  it here (considerin PR size)..

marcgc21 reviewed (changes requested):
Hi @Santosh69, I'm reviewing some stuff. It looks like as an observator it can be possible:

- Press on START THE TEST button. This is something that can not be possible and observator must go to a similar page like the admin or the user where waits other people to be connected.
- Observator also have access to the bottom bar with many functionalities. Indeed he can not press the button, camera, share screen, microphone because right now he can not do that. I recommend to disable those actions.
- Observator has access to "END CALL" button, which is something that he/she should not have access. Because that role is connecting to a session but no have right that allows him/her to finish it.

Until now I found this.

Santosh69 commented:
hey @marcgc21 

> * Press on START THE TEST button. This is something that can not be possible and observator must go to a similar page like the admin or the user where waits other people to be connected.
<img width="1440" height="900" alt="Screenshot 2026-01-22 at 12 36 18 AM" src="https://github.com/user-attachments/assets/b33b9885-53c3-46b0-bf87-da0bb893cb72" />

 **instead this page, observator must be shown this page, right?:**
 
<img width="1440" height="900" alt="Screenshot 2026-01-22 at 12 37 47 AM" src="https://github.com/user-attachments/assets/9076e146-947c-49ce-8a5c-7ee662e9be00" />

> * Observator also have access to the bottom bar with many functionalities. Indeed he can not press the button, camera, share screen, microphone because right now he can not do that. I recommend to disable those actions.
> * Observator has access to "END CALL" button, which is something that he/she should not have access. Because that role is connecting to a session but no have right that allows him/her to finish it.

**Sure, were fixed actually, but got revereted, ig**

Santosh69 committed:
fix(observator): direct observers to waiting page; hide media controls and prevent observers ending session; add implementation guide

Santosh69 commented:
hiii @marcgc21 , @KarinePistili , please review, lmk your feeback, few things to fix and change.. parallely will fix them

marcgc21 commented:
We still have some problems:

- [x] Observator can not hear the moderator (it should be possible)
- [ ] Observator can share screen (should NOT be POSSIBLE)
- [x] Observator do not appear on the participants list  (it should appear)
- [ ] Leave call button should allow the observator leave the call
- [ ] End call button should not appear for the observator

marcgc21 committed:
fix: wip solving a call with one moderator and one observator

marcgc21 committed:
Merge branch 'develop' into feat/user-test-observator-role

marcgc21 committed:
style(lint): fix lint problems

marcgc21 committed:
style(lint): solving lint problems from observator components
